{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-04-13T17:25:43.755455Z",
     "start_time": "2025-04-13T17:25:43.753130Z"
    }
   },
   "source": [
    "import time\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from Bio import Entrez, SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "from entrez import select_phyla\n",
    "from fasta_processing import read_fasta, read_single_fasta, dict_align_to_fasta\n",
    "from data_processing import choose_best_frameshift, save_subset_df_transcripts, find_codon, download_subset_df_datasets"
   ],
   "outputs": [],
   "execution_count": 207
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T10:06:41.768673Z",
     "start_time": "2025-04-13T10:06:41.744552Z"
    }
   },
   "cell_type": "code",
   "source": [
    "column_names = [\"tax_id\", \"org_name\", \"gene_id\", \"current_id\", \"status\", \"symbol\", \"aliases\", \"description\",\n",
    "                \"other_designations\", \"map_location\", \"chromosome\", \"genomic_nucleotide_accession.version\",\n",
    "                \"start_position_on_the_genomic_accession\", \"end_position_on_the_genomic_accession\", \"orientation\",\n",
    "                \"exon_count\", \"to_delete_1\", \"to_delete_2\"]\n",
    "\n",
    "df_cnidaria = pd.read_csv(\"../all_nxf1.txt\", sep=\"\\t\", skiprows=1, names=column_names, index_col=0)\n",
    "df_cnidaria.drop([\"to_delete_1\", \"to_delete_2\"], axis=1, inplace=True)\n",
    "df_cnidaria.head()"
   ],
   "id": "93e9dab343ef426e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                      org_name    gene_id  current_id status symbol  \\\n",
       "tax_id                                                                \n",
       "8407           Rana temporaria  120917577           0   live   NXF1   \n",
       "8364        Xenopus tropicalis     734058           0   live   nxf1   \n",
       "9598           Pan troglodytes     451267           0   live   NXF1   \n",
       "9913                Bos taurus     512136           0   live   NXF1   \n",
       "9615    Canis lupus familiaris     483780           0   live   NXF1   \n",
       "\n",
       "               aliases                  description  \\\n",
       "tax_id                                                \n",
       "8407               NaN  nuclear RNA export factor 1   \n",
       "8364        mex67, tap  nuclear RNA export factor 1   \n",
       "9598    CK820_G0030982  nuclear RNA export factor 1   \n",
       "9913               NaN  nuclear RNA export factor 1   \n",
       "9615               NaN  nuclear RNA export factor 1   \n",
       "\n",
       "                 other_designations  map_location chromosome  \\\n",
       "tax_id                                                         \n",
       "8407    nuclear RNA export factor 1           NaN         11   \n",
       "8364    nuclear RNA export factor 1           NaN          4   \n",
       "9598    nuclear RNA export factor 1           NaN          9   \n",
       "9913    nuclear RNA export factor 1           NaN         29   \n",
       "9615    nuclear RNA export factor 1           NaN         18   \n",
       "\n",
       "       genomic_nucleotide_accession.version  \\\n",
       "tax_id                                        \n",
       "8407                            NC_053499.1   \n",
       "8364                            NC_030680.2   \n",
       "9598                            NC_072407.2   \n",
       "9913                            NC_037356.1   \n",
       "9615                            NC_051822.1   \n",
       "\n",
       "        start_position_on_the_genomic_accession  \\\n",
       "tax_id                                            \n",
       "8407                                120752259.0   \n",
       "8364                                 36493131.0   \n",
       "9598                                 65104305.0   \n",
       "9913                                 41099477.0   \n",
       "9615                                 54936964.0   \n",
       "\n",
       "        end_position_on_the_genomic_accession orientation  exon_count  \n",
       "tax_id                                                                 \n",
       "8407                              120809664.0        plus        22.0  \n",
       "8364                               36532401.0       minus        23.0  \n",
       "9598                               65118449.0       minus        23.0  \n",
       "9913                               41110792.0       minus        22.0  \n",
       "9615                               54947631.0        plus        21.0  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>org_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>current_id</th>\n",
       "      <th>status</th>\n",
       "      <th>symbol</th>\n",
       "      <th>aliases</th>\n",
       "      <th>description</th>\n",
       "      <th>other_designations</th>\n",
       "      <th>map_location</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genomic_nucleotide_accession.version</th>\n",
       "      <th>start_position_on_the_genomic_accession</th>\n",
       "      <th>end_position_on_the_genomic_accession</th>\n",
       "      <th>orientation</th>\n",
       "      <th>exon_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tax_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8407</th>\n",
       "      <td>Rana temporaria</td>\n",
       "      <td>120917577</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>NXF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>NC_053499.1</td>\n",
       "      <td>120752259.0</td>\n",
       "      <td>120809664.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8364</th>\n",
       "      <td>Xenopus tropicalis</td>\n",
       "      <td>734058</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>nxf1</td>\n",
       "      <td>mex67, tap</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>NC_030680.2</td>\n",
       "      <td>36493131.0</td>\n",
       "      <td>36532401.0</td>\n",
       "      <td>minus</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9598</th>\n",
       "      <td>Pan troglodytes</td>\n",
       "      <td>451267</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>NXF1</td>\n",
       "      <td>CK820_G0030982</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9</td>\n",
       "      <td>NC_072407.2</td>\n",
       "      <td>65104305.0</td>\n",
       "      <td>65118449.0</td>\n",
       "      <td>minus</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9913</th>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>512136</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>NXF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29</td>\n",
       "      <td>NC_037356.1</td>\n",
       "      <td>41099477.0</td>\n",
       "      <td>41110792.0</td>\n",
       "      <td>minus</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9615</th>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>483780</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>NXF1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18</td>\n",
       "      <td>NC_051822.1</td>\n",
       "      <td>54936964.0</td>\n",
       "      <td>54947631.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T10:06:41.898280Z",
     "start_time": "2025-04-13T10:06:41.896126Z"
    }
   },
   "cell_type": "code",
   "source": "# cnidaria = select_phyla(df_cnidaria, \"Cnidaria\")",
   "id": "dd07671209454602",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T10:06:41.921670Z",
     "start_time": "2025-04-13T10:06:41.919080Z"
    }
   },
   "cell_type": "code",
   "source": "cnidaria_copy = {'Cnidaria': [50429, 6105, 6087, 45351]}",
   "id": "359784edb61526f7",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T13:53:12.975940Z",
     "start_time": "2025-04-13T13:53:12.972993Z"
    }
   },
   "cell_type": "code",
   "source": "cnidaria_subset = df_cnidaria.loc[cnidaria_copy['Cnidaria']]",
   "id": "719f7bb1f43a415f",
   "outputs": [],
   "execution_count": 147
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T10:06:42.079659Z",
     "start_time": "2025-04-13T10:06:42.064982Z"
    }
   },
   "cell_type": "code",
   "source": "cnidaria_subset",
   "id": "55a73e9fad782bf1",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                      org_name    gene_id  current_id status        symbol  \\\n",
       "tax_id                                                                       \n",
       "50429    Stylophora pistillata  111339613           0   live  LOC111339613   \n",
       "6105         Actinia tenebrosa  116292710           0   live  LOC116292710   \n",
       "6087            Hydra vulgaris  100214173           0   live  LOC100214173   \n",
       "45351   Nematostella vectensis    5510737           0   live    LOC5510737   \n",
       "\n",
       "                          aliases                       description  \\\n",
       "tax_id                                                                \n",
       "50429   AWC38_SpisGene17253, NXF1  nuclear RNA export factor 1-like   \n",
       "6105                          NaN  nuclear RNA export factor 1-like   \n",
       "6087                          NaN       nuclear RNA export factor 1   \n",
       "45351        NEMVEDRAFT_v1g244041       nuclear RNA export factor 1   \n",
       "\n",
       "                                       other_designations  map_location  \\\n",
       "tax_id                                                                    \n",
       "50429   nuclear RNA export factor 1-like|Nuclear RNA e...           NaN   \n",
       "6105                     nuclear RNA export factor 1-like           NaN   \n",
       "6087                          nuclear RNA export factor 1           NaN   \n",
       "45351       nuclear RNA export factor 1|predicted protein           NaN   \n",
       "\n",
       "       chromosome genomic_nucleotide_accession.version  \\\n",
       "tax_id                                                   \n",
       "50429          Un                       NW_019218197.1   \n",
       "6105           Un                       NW_022259381.1   \n",
       "6087           12                          NC_088975.1   \n",
       "45351           7                          NC_064040.1   \n",
       "\n",
       "        start_position_on_the_genomic_accession  \\\n",
       "tax_id                                            \n",
       "50429                                   98177.0   \n",
       "6105                                    16092.0   \n",
       "6087                                   736736.0   \n",
       "45351                                11955766.0   \n",
       "\n",
       "        end_position_on_the_genomic_accession orientation  exon_count  \n",
       "tax_id                                                                 \n",
       "50429                                131366.0        plus        21.0  \n",
       "6105                                  33206.0        plus        24.0  \n",
       "6087                                 739315.0       minus         2.0  \n",
       "45351                              11973360.0        plus        24.0  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>org_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>current_id</th>\n",
       "      <th>status</th>\n",
       "      <th>symbol</th>\n",
       "      <th>aliases</th>\n",
       "      <th>description</th>\n",
       "      <th>other_designations</th>\n",
       "      <th>map_location</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genomic_nucleotide_accession.version</th>\n",
       "      <th>start_position_on_the_genomic_accession</th>\n",
       "      <th>end_position_on_the_genomic_accession</th>\n",
       "      <th>orientation</th>\n",
       "      <th>exon_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tax_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50429</th>\n",
       "      <td>Stylophora pistillata</td>\n",
       "      <td>111339613</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>LOC111339613</td>\n",
       "      <td>AWC38_SpisGene17253, NXF1</td>\n",
       "      <td>nuclear RNA export factor 1-like</td>\n",
       "      <td>nuclear RNA export factor 1-like|Nuclear RNA e...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_019218197.1</td>\n",
       "      <td>98177.0</td>\n",
       "      <td>131366.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6105</th>\n",
       "      <td>Actinia tenebrosa</td>\n",
       "      <td>116292710</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>LOC116292710</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nuclear RNA export factor 1-like</td>\n",
       "      <td>nuclear RNA export factor 1-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_022259381.1</td>\n",
       "      <td>16092.0</td>\n",
       "      <td>33206.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6087</th>\n",
       "      <td>Hydra vulgaris</td>\n",
       "      <td>100214173</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>LOC100214173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>NC_088975.1</td>\n",
       "      <td>736736.0</td>\n",
       "      <td>739315.0</td>\n",
       "      <td>minus</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45351</th>\n",
       "      <td>Nematostella vectensis</td>\n",
       "      <td>5510737</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>LOC5510737</td>\n",
       "      <td>NEMVEDRAFT_v1g244041</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>nuclear RNA export factor 1|predicted protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>NC_064040.1</td>\n",
       "      <td>11955766.0</td>\n",
       "      <td>11973360.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T10:06:42.140920Z",
     "start_time": "2025-04-13T10:06:42.138851Z"
    }
   },
   "cell_type": "code",
   "source": "# download_subset_df_datasets(cnidaria_subset)",
   "id": "5eb211c60cb9d9c",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T17:18:43.578218Z",
     "start_time": "2025-04-13T17:18:43.572824Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def download_gene_gb(org_names: list) -> None:\n",
    "    # Obtain mRNA Accession and ranges of the gene\n",
    "    for org_name in org_names:\n",
    "        with open(f\"../Datasets/{org_name}/ncbi_dataset/data/gene.fna\") as infile:\n",
    "            line = infile.readline().rstrip()\n",
    "            gene_acc = line.split(\":\")[0][1:]\n",
    "            gene_ranges = line.split()[0].split(\":\")[1]\n",
    "            gene_range_1 = gene_ranges.split(\"-\")[0]\n",
    "            gene_range_2 = gene_ranges.split(\"-\")[1]\n",
    "\n",
    "        if \"c\" in gene_range_1:\n",
    "            gene_range_1 = gene_range_1.replace(\"c\", \"\")\n",
    "            strand = 2\n",
    "        else:\n",
    "            strand = 1\n",
    "        gene_range_1, gene_range_2 = int(gene_range_1), int(gene_range_2)\n",
    "\n",
    "        # Obtain and download mRNA in GenBank format\n",
    "        stream = Entrez.efetch(db=\"nucleotide\", id=gene_acc, idtype=\"acc\",\n",
    "                               seq_start=gene_range_1, seq_stop=gene_range_2, strand=strand,\n",
    "                               rettype=\"gb\", retmode=\"text\")\n",
    "        with open(f\"../Datasets/{org_name}/ncbi_dataset/data/gene.gb\", \"w\") as outfile:\n",
    "            outfile.write(stream.read())\n",
    "\n",
    "        print(f\"{org_name}.gb has been downloaded\")\n",
    "\n",
    "        time.sleep(0.333333334)"
   ],
   "id": "44a874cea0c8e073",
   "outputs": [],
   "execution_count": 204
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T17:18:56.434686Z",
     "start_time": "2025-04-13T17:18:48.155688Z"
    }
   },
   "cell_type": "code",
   "source": "download_gene_gb([\"stylophora_pistillata\", \"nematostella_vectensis\", \"actinia_tenebrosa\", \"hydra_vulgaris\"])",
   "id": "a329b8dbb4365d6c",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stylophora_pistillata.gb has been downloaded\n",
      "nematostella_vectensis.gb has been downloaded\n",
      "actinia_tenebrosa.gb has been downloaded\n",
      "hydra_vulgaris.gb has been downloaded\n"
     ]
    }
   ],
   "execution_count": 205
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T18:42:32.780586Z",
     "start_time": "2025-04-13T18:42:32.776231Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def parse_exon_ranges(org_names: list, feature_type: str = \"mRNA\") -> dict:\n",
    "    exons_dict = {}\n",
    "\n",
    "    feature_types = [\"mRNA\", \"CDS\"]\n",
    "    if feature_type not in feature_types:\n",
    "        raise Exception(f\"Feature type {feature_type} is not supported\")\n",
    "\n",
    "    for org_name in org_names:\n",
    "        exons_range = {}\n",
    "        records = SeqIO.parse(f\"../Datasets/{org_name}/ncbi_dataset/data/gene.gb\", \"genbank\")\n",
    "        for record in records:\n",
    "            for feature in record.features:\n",
    "                if feature.type == feature_type:\n",
    "                    for part_i, part in enumerate(feature.location.parts):\n",
    "                        start, end = int(part.start), int(part.end)\n",
    "                        exons_range[part_i] = [start, end]\n",
    "        exons_dict[org_name] = exons_range\n",
    "\n",
    "    return exons_dict"
   ],
   "id": "c5b1b0214e54a5de",
   "outputs": [],
   "execution_count": 291
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T18:44:27.840321Z",
     "start_time": "2025-04-13T18:44:27.837497Z"
    }
   },
   "cell_type": "code",
   "source": "org_names = [\"stylophora_pistillata\", \"nematostella_vectensis\", \"actinia_tenebrosa\", \"hydra_vulgaris\"]",
   "id": "4c2e64e759fdf2d9",
   "outputs": [],
   "execution_count": 297
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T18:44:28.856934Z",
     "start_time": "2025-04-13T18:44:28.846124Z"
    }
   },
   "cell_type": "code",
   "source": "parse_exon_ranges(org_names, feature_type=\"mRNA\")",
   "id": "3901caed6a209c11",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'stylophora_pistillata': {0: [0, 243],\n",
       "  1: [882, 1092],\n",
       "  2: [18604, 18771],\n",
       "  3: [19302, 19441],\n",
       "  4: [20748, 20852],\n",
       "  5: [21589, 21659],\n",
       "  6: [22484, 22573],\n",
       "  7: [23168, 23276],\n",
       "  8: [23896, 24012],\n",
       "  9: [24495, 24546],\n",
       "  10: [26302, 26363],\n",
       "  11: [26884, 26943],\n",
       "  12: [27066, 27128],\n",
       "  13: [28278, 28394],\n",
       "  14: [28638, 28681],\n",
       "  15: [29443, 29507],\n",
       "  16: [30474, 30532],\n",
       "  17: [30851, 30938],\n",
       "  18: [31526, 31588],\n",
       "  19: [31927, 31988],\n",
       "  20: [32669, 33190]},\n",
       " 'nematostella_vectensis': {0: [0, 218],\n",
       "  1: [1286, 1536],\n",
       "  2: [2152, 2300],\n",
       "  3: [2733, 2817],\n",
       "  4: [3225, 3330],\n",
       "  5: [4828, 4939],\n",
       "  6: [5047, 5117],\n",
       "  7: [5635, 5724],\n",
       "  8: [6199, 6307],\n",
       "  9: [7012, 7128],\n",
       "  10: [8119, 8156],\n",
       "  11: [8298, 8367],\n",
       "  12: [8531, 8587],\n",
       "  13: [10051, 10112],\n",
       "  14: [10520, 10576],\n",
       "  15: [11031, 11093],\n",
       "  16: [11664, 11780],\n",
       "  17: [12355, 12398],\n",
       "  18: [12862, 12926],\n",
       "  19: [13499, 13557],\n",
       "  20: [13681, 13819],\n",
       "  21: [14909, 14974],\n",
       "  22: [16026, 16087],\n",
       "  23: [17169, 17595]},\n",
       " 'actinia_tenebrosa': {0: [0, 183],\n",
       "  1: [793, 1022],\n",
       "  2: [1560, 1693],\n",
       "  3: [1942, 2026],\n",
       "  4: [3305, 3410],\n",
       "  5: [4085, 4208],\n",
       "  6: [4273, 4343],\n",
       "  7: [4434, 4523],\n",
       "  8: [5748, 5856],\n",
       "  9: [6442, 6558],\n",
       "  10: [6731, 6768],\n",
       "  11: [6862, 6931],\n",
       "  12: [8209, 8265],\n",
       "  13: [9860, 9921],\n",
       "  14: [10484, 10540],\n",
       "  15: [11330, 11392],\n",
       "  16: [12381, 12497],\n",
       "  17: [13190, 13233],\n",
       "  18: [13317, 13381],\n",
       "  19: [13623, 13681],\n",
       "  20: [14380, 14545],\n",
       "  21: [15090, 15152],\n",
       "  22: [15659, 15720],\n",
       "  23: [16396, 17115]},\n",
       " 'hydra_vulgaris': {0: [0, 351], 1: [751, 2580]}}"
      ]
     },
     "execution_count": 298,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 298
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T17:34:09.920514Z",
     "start_time": "2025-04-13T17:34:09.907482Z"
    }
   },
   "cell_type": "code",
   "source": [
    "records = SeqIO.parse(f\"../Datasets/{org_name}/ncbi_dataset/data/gene.gb\", \"genbank\")\n",
    "\n",
    "for record in records:\n"
   ],
   "id": "15fb66203f46207d",
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "list indices must be integers or slices, not str",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mTypeError\u001B[0m                                 Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[231], line 4\u001B[0m\n\u001B[1;32m      1\u001B[0m records \u001B[38;5;241m=\u001B[39m SeqIO\u001B[38;5;241m.\u001B[39mparse(\u001B[38;5;124mf\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124m../Datasets/\u001B[39m\u001B[38;5;132;01m{\u001B[39;00morg_name\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m/ncbi_dataset/data/gene.gb\u001B[39m\u001B[38;5;124m\"\u001B[39m, \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mgenbank\u001B[39m\u001B[38;5;124m\"\u001B[39m)\n\u001B[1;32m      3\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m record \u001B[38;5;129;01min\u001B[39;00m records:\n\u001B[0;32m----> 4\u001B[0m     \u001B[38;5;28mprint\u001B[39m(\u001B[43mrecord\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mfeatures\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mmRNA\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m]\u001B[49m)\n",
      "\u001B[0;31mTypeError\u001B[0m: list indices must be integers or slices, not str"
     ]
    }
   ],
   "execution_count": 231
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "bf71a35da15d48e2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "278fdcc99e6fce13"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "3b012da47ac764c6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "6ffafd6858ff3bb9"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "17c2f083d0394ee7"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T14:08:40.941605Z",
     "start_time": "2025-04-13T14:08:40.937494Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def download_exon_ranges_auto(org_names: list) -> dict:\n",
    "    exons_dict = {}\n",
    "\n",
    "    for org_name in org_names:\n",
    "        exons = []\n",
    "        record = False\n",
    "\n",
    "        with open(f\"../Datasets/{org_name}/ncbi_dataset/data/rna.gb\") as infile:\n",
    "            for line in infile.readlines():\n",
    "                line = line.rstrip()\n",
    "                if line.startswith(\"PRIMARY\"):\n",
    "                    record = True\n",
    "                    continue\n",
    "                if record:\n",
    "                    if line.startswith(\"FEATURES\"):\n",
    "                        break\n",
    "                    else:\n",
    "                        exons.append(\" \".join(line.split()))\n",
    "\n",
    "        exons_range = []\n",
    "        for exon in exons:\n",
    "            exon_range = exon.split()[0]\n",
    "            exon_range_0, exon_range_1 = exon_range.split(\"-\")[0], exon_range.split(\"-\")[1]\n",
    "            exons_range.append([int(exon_range_0), int(exon_range_1)])\n",
    "        exons_dict[org_name] = exons_range\n",
    "\n",
    "    return exons_dict"
   ],
   "id": "3f8bf10d60f30891",
   "outputs": [],
   "execution_count": 166
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "with open(f\"../Datasets/{org_name}/ncbi_dataset/data/rna.gb\") as infile:\n",
    "    for line_i, line in enumerate(infile.readlines()):\n",
    "        line = line.rstrip()\n",
    "        print(f\"line {line_i}: {line}\")\n",
    "        time.sleep(2)\n",
    "        print(\"Enter the range of exons via space:\")\n",
    "        custom_range = [int(i) for i in input().split()]\n",
    "\n",
    "        with open(f\"../Datasets/{org_name}/ncbi_dataset/data/rna.gb\") as infile:\n",
    "            for line_i, line in enumerate(infile.readlines()):\n",
    "                line = line.rstrip()\n",
    "                if line_i == custom_range[0]:\n",
    "                    record = True\n",
    "                    continue\n",
    "                if record:\n",
    "                    if line_i == custom_range[1]:\n",
    "                        break\n",
    "                    else:\n",
    "                        exons.append(\" \".join(line.split()))"
   ],
   "id": "6937c83fcc4cbae7"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "org_names = [\"stylophora_pistillata\", \"nematostella_vectensis\"]",
   "id": "c3ab537a12224652"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T14:08:53.137816Z",
     "start_time": "2025-04-13T14:08:53.132993Z"
    }
   },
   "cell_type": "code",
   "source": "download_exon_ranges_auto(org_names)",
   "id": "48e81c4dc5ca93ae",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'stylophora_pistillata': [[1, 243],\n",
       "  [244, 453],\n",
       "  [454, 620],\n",
       "  [621, 759],\n",
       "  [760, 976],\n",
       "  [977, 1080],\n",
       "  [1081, 1150],\n",
       "  [1151, 1239],\n",
       "  [1240, 1347],\n",
       "  [1348, 1463],\n",
       "  [1464, 1514],\n",
       "  [1515, 1575],\n",
       "  [1576, 1634],\n",
       "  [1635, 1696],\n",
       "  [1697, 1812],\n",
       "  [1813, 1855],\n",
       "  [1856, 1919],\n",
       "  [1920, 1977],\n",
       "  [1978, 2064],\n",
       "  [2065, 2126],\n",
       "  [2127, 2187],\n",
       "  [2188, 2708]]}"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 168
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T13:25:29.463786Z",
     "start_time": "2025-04-13T13:25:29.460877Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "id": "a795b58162272db3",
   "outputs": [],
   "execution_count": 139
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# почему у меня получилось 22 экзона, а в табличке написано, что 21?\n",
    "# возможно дело в том, что у нас есть GARY01005553.1 == UTR"
   ],
   "id": "e20d3bfb7df2c12a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "11e44d0f942241e3"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "org_name = \"stylophora_pistillata\"\n",
    "stylophora_rna = read_single_fasta(f\"../Datasets/{org_name}/ncbi_dataset/data/rna.fasta\")\n"
   ],
   "id": "9c353f042c3e894"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T11:38:23.813911Z",
     "start_time": "2025-04-13T11:38:23.810316Z"
    }
   },
   "cell_type": "code",
   "source": "cnidaria_subset[\"exons\"] = [1, 2, 3, 4]",
   "id": "4d337e4156e77efb",
   "outputs": [],
   "execution_count": 120
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T13:15:04.827272Z",
     "start_time": "2025-04-13T13:15:04.822961Z"
    }
   },
   "cell_type": "code",
   "source": "cnidaria_subset.loc[cnidaria_subset[cnidaria_subset[\"gene_id\"] == 111339613].index, \"exons\"] = 1337",
   "id": "f6718b1e6a76edc4",
   "outputs": [],
   "execution_count": 126
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-13T13:15:08.056835Z",
     "start_time": "2025-04-13T13:15:08.045402Z"
    }
   },
   "cell_type": "code",
   "source": "cnidaria_subset",
   "id": "d0d208b9e87cde30",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                      org_name    gene_id  current_id status        symbol  \\\n",
       "tax_id                                                                       \n",
       "50429    Stylophora pistillata  111339613           0   live  LOC111339613   \n",
       "6105         Actinia tenebrosa  116292710           0   live  LOC116292710   \n",
       "6087            Hydra vulgaris  100214173           0   live  LOC100214173   \n",
       "45351   Nematostella vectensis    5510737           0   live    LOC5510737   \n",
       "\n",
       "                          aliases                       description  \\\n",
       "tax_id                                                                \n",
       "50429   AWC38_SpisGene17253, NXF1  nuclear RNA export factor 1-like   \n",
       "6105                          NaN  nuclear RNA export factor 1-like   \n",
       "6087                          NaN       nuclear RNA export factor 1   \n",
       "45351        NEMVEDRAFT_v1g244041       nuclear RNA export factor 1   \n",
       "\n",
       "                                       other_designations  map_location  \\\n",
       "tax_id                                                                    \n",
       "50429   nuclear RNA export factor 1-like|Nuclear RNA e...           NaN   \n",
       "6105                     nuclear RNA export factor 1-like           NaN   \n",
       "6087                          nuclear RNA export factor 1           NaN   \n",
       "45351       nuclear RNA export factor 1|predicted protein           NaN   \n",
       "\n",
       "       chromosome genomic_nucleotide_accession.version  \\\n",
       "tax_id                                                   \n",
       "50429          Un                       NW_019218197.1   \n",
       "6105           Un                       NW_022259381.1   \n",
       "6087           12                          NC_088975.1   \n",
       "45351           7                          NC_064040.1   \n",
       "\n",
       "        start_position_on_the_genomic_accession  \\\n",
       "tax_id                                            \n",
       "50429                                   98177.0   \n",
       "6105                                    16092.0   \n",
       "6087                                   736736.0   \n",
       "45351                                11955766.0   \n",
       "\n",
       "        end_position_on_the_genomic_accession orientation  exon_count  exons  \n",
       "tax_id                                                                        \n",
       "50429                                131366.0        plus        21.0   1337  \n",
       "6105                                  33206.0        plus        24.0      2  \n",
       "6087                                 739315.0       minus         2.0      3  \n",
       "45351                              11973360.0        plus        24.0      4  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>org_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>current_id</th>\n",
       "      <th>status</th>\n",
       "      <th>symbol</th>\n",
       "      <th>aliases</th>\n",
       "      <th>description</th>\n",
       "      <th>other_designations</th>\n",
       "      <th>map_location</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genomic_nucleotide_accession.version</th>\n",
       "      <th>start_position_on_the_genomic_accession</th>\n",
       "      <th>end_position_on_the_genomic_accession</th>\n",
       "      <th>orientation</th>\n",
       "      <th>exon_count</th>\n",
       "      <th>exons</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tax_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50429</th>\n",
       "      <td>Stylophora pistillata</td>\n",
       "      <td>111339613</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>LOC111339613</td>\n",
       "      <td>AWC38_SpisGene17253, NXF1</td>\n",
       "      <td>nuclear RNA export factor 1-like</td>\n",
       "      <td>nuclear RNA export factor 1-like|Nuclear RNA e...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_019218197.1</td>\n",
       "      <td>98177.0</td>\n",
       "      <td>131366.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6105</th>\n",
       "      <td>Actinia tenebrosa</td>\n",
       "      <td>116292710</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>LOC116292710</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nuclear RNA export factor 1-like</td>\n",
       "      <td>nuclear RNA export factor 1-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Un</td>\n",
       "      <td>NW_022259381.1</td>\n",
       "      <td>16092.0</td>\n",
       "      <td>33206.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>24.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6087</th>\n",
       "      <td>Hydra vulgaris</td>\n",
       "      <td>100214173</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>LOC100214173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>NC_088975.1</td>\n",
       "      <td>736736.0</td>\n",
       "      <td>739315.0</td>\n",
       "      <td>minus</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45351</th>\n",
       "      <td>Nematostella vectensis</td>\n",
       "      <td>5510737</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>LOC5510737</td>\n",
       "      <td>NEMVEDRAFT_v1g244041</td>\n",
       "      <td>nuclear RNA export factor 1</td>\n",
       "      <td>nuclear RNA export factor 1|predicted protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>NC_064040.1</td>\n",
       "      <td>11955766.0</td>\n",
       "      <td>11973360.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>24.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 127
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
